File: examples/laser_tag/src/ajan_pomdp_planner/cpp/ajan_particleupperbound.cpp
//
// Created by bhuvanesh on 01.08.23.
// JNI Interface for Particle UpperBound and Belief UpperBound.
// See: examples/laser_tag/include/unified_pomdp_planner/tag_sp_particleupperbound.h
//

#include "ajan_particleupperbound.h"
namespace despot {
using namespace despot;

    AjanParticleUpperBound::AjanParticleUpperBound(const DSPOMDP *model):
    tag_model_(model) { // Changed from AjanAgent to DSPOMDP and it works well
        // TODO: Implement AjanParticleUpperBound constructor to call JNI
        // store the distance between rob and opponent in value_[state.state_id]
        /**
         * This is implementing both particle and belief upperbound. So, try to separate them.
        **/
    }

    double AjanParticleUpperBound::Value(const despot::State &s) const {
        // TODO: Implement AjanParticleUpperBound::Value Function for Particle to call JNI.
        // Return precomputed value_[state.state_id];
        jobject javaState = AjanHelper::toJavaState(s);
        return AjanHelper::getEnv()->CallDoubleMethod(javaReferenceObject, AjanHelper::getMethodID(AJAN_PARTICLE_UPPER_BOUND,Value_), javaState);
    }

}

File: examples/laser_tag/src/ajan_pomdp_planner/java/src/main/java/de/dfki/asr/ajan/pluginsystem/mdpplugin/utils/POMDP/DESPOT/implementation/AjanParticleUpperBound.java
package de.dfki.asr.ajan.pluginsystem.mdpplugin.utils.POMDP.DESPOT.implementation;

import de.dfki.asr.ajan.pluginsystem.mdpplugin.utils.POMDP.DESPOT.interface_.ParticleUpperBound;
import de.dfki.asr.ajan.pluginsystem.mdpplugin.utils.POMDP.DESPOT.interface_.State;

import java.util.Vector;

public class AjanParticleUpperBound extends ParticleUpperBound {
    public AjanAgent agent_model;
    public Vector<Double> value_;

    AjanParticleUpperBound(AjanAgent model) {
        agent_model = model;
        // TODO: Implement AjanParticleUpperBound constructor to call JNI
        // store the distance between rob and opponent in value_[state.state_id]
        /*
         * This is implementing both particle and belief upperbound. So, try to separate them.
         **/
    }
    AjanParticleUpperBound(long cppPointer){

    }
    @Override
    protected double Value(State state) {
        // TODO: Implement AjanParticleUpperBound::Value Function for Particle to call JNI.
        // Return precomputed value_[state.state_id];
        return 0.0;
    }

}
